File: scripts/check.py
"""Validate skill links, Python syntax and the public repository boundary."""

from pathlib import Path
import re
import subprocess

ROOT = Path(__file__).resolve().parents[1]
skill = ROOT / "skills/japanese-learning"
text = (skill / "SKILL.md").read_text(encoding="utf-8")
assert text.startswith("---\nname: japanese-learning\ndescription: ")
assert text.split("---", 2)[1].strip().splitlines()[1].removeprefix("description:").strip()
for path in ROOT.rglob("*.md"):
    if ".git" in path.parts:
        continue
    for link in re.findall(r"\]\(([^)]+)\)", path.read_text(encoding="utf-8")):
        if "://" not in link and not link.startswith("#"):
            assert (path.parent / link.split("#", 1)[0]).exists(), (path, link)
for path in ROOT.rglob("*.py"):
    compile(path.read_text(encoding="utf-8"), str(path), "exec")
# Check staged and untracked, non-ignored files before a first commit as well as in CI.
files = subprocess.check_output(["git", "ls-files", "--cached", "--others", "--exclude-standard", "-z"], cwd=ROOT).decode().split("\0")
for name in filter(None, files):
    path = Path(name)
    assert not {"private", "state", "events", "artifacts"}.intersection(path.parts), name
    assert path.name != "profile.json" and path.suffix not in {".pdf", ".png", ".jpg", ".docx"}, name
    content = (ROOT / path).read_text(encoding="utf-8")
    assert not re.search(r"/(?:Users|home)/[^/\s]+/", content), f"Personal path in {name}"
print("Skill structure, local links, Python syntax and public file boundary checked.")

File: scripts/check.mjs
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { basename, dirname, extname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = fileURLToPath(new URL("../", import.meta.url));
for (const name of readdirSync(join(root, "skills"))) {
  const skill = readFileSync(join(root, "skills", name, "SKILL.md"), "utf8");
  assert(skill.startsWith(`---\nname: ${name}\ndescription: `));
  assert(/^description: .+$/m.test(skill.split("---")[1]));
}
// Include non-ignored new files so the same check also works before committing.
const files = execFileSync(
  "git",
  ["ls-files", "--cached", "--others", "--exclude-standard", "-z"],
  { cwd: root, encoding: "utf8" },
)
  .split("\0")
  .filter(Boolean);
for (const name of new Set(files)) {
  const path = join(root, name);
  if (!existsSync(path)) continue;
  assert(
    !name.split("/").some((part) => ["private", "state", "events", "artifacts"].includes(part)),
    name,
  );
  assert(
    basename(name) !== "profile.json" && ![".pdf", ".png", ".jpg", ".docx"].includes(extname(name)),
    name,
  );
  const content = readFileSync(path, "utf8");
  assert(!/\/(?:Users|home)\/[^/\s]+\//.test(content), `Personal path in ${name}`);
  if (extname(name) === ".md") {
    for (const [, link] of content.matchAll(/\]\(([^)]+)\)/g)) {
      if (!link.includes("://") && !link.startsWith("#")) {
        assert(existsSync(resolve(dirname(path), link.split("#")[0])), `${name}: ${link}`);
      }
    }
  }
  if (extname(name) === ".mjs") execFileSync(process.execPath, ["--check", path]);
}
console.log("Skill structure, local links, JavaScript syntax and public file boundary checked.");
